### PR TITLE
[allms] Upgrade to Gpt-4o

### PR DESCRIPTION
**Pull request by kamillitman**
- [ ] Added support of GPT-4o
- [ ] Added support of Assistants Beta V2
- [ ] Updated rate limits
- [ ] Version bump

### CODE REVIEW

File: src/assistant.rs [hunk 1]
```diff
@@ -63,25 +76,27 @@ impl OpenAIAssistant {
      * This function creates an Assistant and updates the ID of the OpenAIAssistant struct
      */
     async fn create_assistant(&mut self) -> Result<()> {
-        //Get the API url
-        let assistant_url = "https://api.openai.com/v1/assistants";
+        //Get the assistant API url
```
Comment (darioalessandro) [inline src/assistant.rs:79]: Does this break assistant v1?

File: src/assistant.rs [hunk 2]
```diff
@@ -241,15 +253,18 @@ impl OpenAIAssistant {
      * This function creates a Thread and updates the thread_id of the OpenAIAssistant struct
      */
     async fn create_thread(&mut self, body: &serde_json::Value) -> Result<()> {
-        let thread_url = "https://api.openai.com/v1/threads";
+        //Get version-specific URL
```
Comment (darioalessandro) [inline src/assistant.rs:256]: does this break the previous api?

File: src/models.rs [hunk 3]
```diff
@@ -283,28 +290,32 @@ impl OpenAIModels {
...
             OpenAIModels::Gpt4 => RateLimit {
-                tpm: 10_000,
-                rpm: 200,
+                tpm: 300_000,
+                rpm: 10_000,
             },
             OpenAIModels::Gpt4Turbo => RateLimit {
-                tpm: 10_000,
-                rpm: 200,
+                tpm: 2_000_000,
```
Comment (darioalessandro) [inline src/models.rs:309]: wow this is a significant bump!!

**Review by darioalessandro — APPROVED** [review]

File: src/assistant.rs (hunk 1 above)
Comment (kamillitman) [reply src/assistant.rs:79]: This does not but the constructor certainly does. Let me address that.

File: src/models.rs (hunk 3 above)
Comment (kamillitman) [reply src/models.rs:309]: Yeah, I was pretty shocked too, but that's what the documentation now shows

File: src/assistant.rs [hunk 4]
```diff
@@ -33,55 +39,59 @@ pub struct OpenAIAssistant {
     instructions: String,
     debug: bool,
     api_key: String,
+    version: OpenAIAssistantVersion,
 }
 
 impl OpenAIAssistant {
     //Constructor
     pub async fn new(model: OpenAIModels, open_ai_key: &str, debug: bool) -> Result<Self> {
```
Comment (kamillitman) [inline src/assistant.rs:47]: @darioalessandro, now there is no change to the constructor so this should be backwards compatible.

The constructor defaults to V1, so everyone that uses it should still be good.

The constructor no longer needs to be async, but I left it that way to preserve backwards compatibility.

**Review by darioalessandro — APPROVED** [review]

**Pull request merged**